docs/src/userguide/extension_types.rst

```diff
@@ -532,6 +532,100 @@ statically sized freelist of ``N`` instances for a given type.  Example::
 
 .. _making_extension_types_weak_referenceable:
 
+.. _existing-pointers-instantiation:
+
+Instantiation from existing C/C++ pointers
+===========================================
+
+It is quite common to want to instantiate an extension class from an existing
+(pointer to a) data structure, often as returned by external C/C++ functions.
+
+As extension classes can only accept Python objects as arguments in their
+contructors, this necessitates the use of factory functions. For example, ::
+
+    from libc.stdlib cimport malloc, free
+
+    # Example C struct
+    ctypedef struct my_c_struct:
+        int a
+        int b
+
+
+    cdef class WrapperClass:
+        """A wrapper class for a C/C++ data structure"""
+        cdef my_c_struct *_ptr
+        cdef bint ptr_owner
+
+        def __cinit__(self):
+            # On cinit, do not create new structure but set pointer to NULL
+            self._ptr = NULL
+            self.ptr_owner = True
+
+        def __dealloc__(self):
+            # De-allocate if not null and flag is set
+            if self._ptr is not NULL and self.ptr_owner is True:
+                free(self._ptr)
+                self._ptr = NULL
+
+        # Extension class properties
+        @property
+        def a(self):
+            return self._ptr.a if self._ptr is not NULL else None
+
+        @property
+        def b(self):
+            return self._ptr.b if self._ptr is not NULL else None
+
+        @staticmethod
+        cdef WrapperClass from_ptr(my_c_struct *_ptr):
+            """Factory function to create WrapperClass objects from
+            given my_c_struct pointer"""
+            # Call to __new__ bypasses __init__ constructor
+            cdef WrapperClass wrapper = WrapperClass.__new__(WrapperClass)
+            wrapper._ptr = _ptr
```

Comment:
Here, the wrapper should not take ownership of the pointer (i.e. not free it). Alternatively, allow passing an explicit flag argument that transfers the ownership into the wrapper, so that it calls `free()` on dealloc.

Reply:
Added flag for pointer ownership, changed ``new_struct`` to make itself owner.